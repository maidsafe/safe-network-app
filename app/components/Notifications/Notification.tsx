import React, { Component } from 'react';
import {
    Grid,
    Paper,
    Box,
    Avatar,
    Typography,
    Button,
    Divider
} from '@material-ui/core';

import { Banner } from 'material-ui-banner';

// Material-ui Icons
import SignalWifiOffIcon from '@material-ui/icons/SignalWifiOff';
import WarningIcon from '@material-ui/icons/Warning';
import InfoIcon from '@material-ui/icons/Info';
import LockIcon from '@material-ui/icons/Lock';
import DiscFullIcon from '@material-ui/icons/DiscFull';
import LoopIcon from '@material-ui/icons/Loop';
import SignalWifi4BarIcon from '@material-ui/icons/SignalWifi4Bar';

import styles from './Notification.css';

interface Props {
    latestNotification: any;
    acceptNotification: any;
    denyNotification: any;
}

export class Notification extends React.PureComponent<Props> {
    render() {
        const {
            latestNotification,
            acceptNotification,
            denyNotification
        } = this.props;

        const components = {
            SignalWifi4BarIcon,
            SignalWifiOffIcon,
            WarningIcon,
            InfoIcon,
            LockIcon,
            DiscFullIcon,
            LoopIcon
        };

        const handleOnAccept = () => {
            acceptNotification( latestNotification );
        };

        const handleOnDeny = () => {
            denyNotification( latestNotification );
        };

        const NotificationIcon =
            components[latestNotification.icon || 'WarningIcon'];

        return (
            <Box className={styles.Base}>
                <Paper elevation={0} className={styles.Container}>
                    <Grid container wrap="nowrap" alignItems="center">
                        <Grid item>
                            <Avatar
                                className={styles.Icon}
                                aria-label="NotificationIcon"
                            >
                                <NotificationIcon
                                    color="action"
                                    fontSize="inherit"
                                />
                            </Avatar>
                        </Grid>
                        <Grid item>
                            <Typography
                                aria-label="NotificationTitle"
                                variant="body2"
                                className={styles.Title}
                            >
                                {latestNotification.title}
                            </Typography>
                        </Grid>
                    </Grid>
                    <Grid
                        container
                        justify="flex-end"
                        className={styles.Actions}
                    >
                        {latestNotification.acceptText && (
                            <Grid item>
                                <Button
                                    color="primary"
                                    aria-label="AcceptNotification"
                                    className={styles.ActionButton}
                                    onClick={handleOnAccept}
                                >
                                    {latestNotification.acceptText}
                                </Button>
                            </Grid>
                        )}
                        <Grid item>
                            <Button
                                aria-label="DenyNotification"
                                color="primary"
                                className={styles.ActionButton}
                                onClick={handleOnDeny}
                            >
                                {latestNotification.denyText}
                            </Button>
                        </Grid>
                    </Grid>
                </Paper>
                <Divider />
            </Box>
        );
    }
}
