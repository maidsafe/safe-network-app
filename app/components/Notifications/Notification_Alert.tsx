import React, { Component } from 'react';
import { logger } from '$Logger';

const { dialog } = require( 'electron' ).remote;

interface Props {
    notificationCheckBox: boolean;
    latestNotification: Record<string, any>;
    acceptNotification: any;
    denyNotification: any;
    toggleCheckBox: any;
}

export class NotificationAlert extends React.PureComponent<Props> {
    render() {
        const {
            latestNotification,
            toggleCheckBox,
            acceptNotification,
            denyNotification,
            notificationCheckBox
        } = this.props;

        if (
            latestNotification.type === 'CLEARNET_WARNING_ALERT' &&
            notificationCheckBox === true
        ) {
            return <div />;
        }

        const handleOnAccept = () => {
            acceptNotification( { ...latestNotification } );
        };

        const handleOnDeny = () => {
            denyNotification( { ...latestNotification } );
        };

        const options = {
            icon: './resources/icons/32x32.png',
            title: latestNotification.title,
            message: latestNotification.message,
            buttons: latestNotification.buttons,
            checkboxLabel: latestNotification.checkboxLabel
                ? latestNotification.checkboxLabel
                : null
        };

        const responseHandler = ( response, checkboxChecked ) => {
            const buttonLength = options.buttons.length - 1;

            switch ( response ) {
                case 0:
                    // eslint-disable-next-line no-unused-expressions
                    buttonLength === 1
                        ? handleOnDeny()
                        : logger.error( 'Invalid Action Click' );
                    break;
                case 1:
                    // eslint-disable-next-line no-unused-expressions
                    buttonLength === 1 ? handleOnAccept() : handleOnDeny();
                    break;
                case 2:
                    handleOnAccept();
                    break;
                default:
                    logger.error( 'Invalid Button Click' );
            }

            if ( options.checkboxLabel ) {
                toggleCheckBox( { checkboxChecked } );
            }
        };

        const alert = dialog.showMessageBox(
            null,
            // @ts-ignore
            options,
            ( response, checkboxChecked ) =>
                responseHandler( response, checkboxChecked )
        );
        return <div />;
    }
}
