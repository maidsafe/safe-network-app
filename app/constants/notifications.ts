import { I18n } from 'react-redux-i18n';

const prepareNotification = ( title, acceptText, denyText, otherProperties ) => {
    const id: string = Math.random().toString( 36 );

    return {
        id,
        title,
        acceptText,
        denyText,
        ...otherProperties
    };
};

export const notificationTypes = {
    NO_INTERNET: ( appId ) => {
        const title = I18n.t( 'notifications.title.no_internet' );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.resume' );
        const denyText = I18n.t( 'notifications.buttons.denyText.dismiss' );
        const otherProperties = {
            type: 'NO_INTERNET',
            icon: 'SignalWifiOffIcon',
            priority: 'HIGH',
            notificationType: 'Native',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    CLOSE_APP: ( appId, appName, application ) => {
        const title = I18n.t( 'notifications.title.close_app', { appName } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.try_again' );
        const denyText = I18n.t(
            'notifications.buttons.denyText.cancel_install'
        );
        if ( application === null ) {
            throw new Error( 'Need To Pass application' );
        }

        const otherProperties = {
            type: 'CLOSE_APP',
            icon: 'InfoIcon',
            notificationType: 'Native',
            priority: 'LOW',
            application,
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    CLOSE_APP_ALERT: ( appId, appName, application ) => {
        const title = I18n.t( 'notifications.title.close_app', { appName } );
        const message = I18n.t( 'notifications.message.close_app', { appName } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.try_again' );
        const denyText = I18n.t(
            'notifications.buttons.denyText.cancel_install'
        );
        const buttons = [denyText, acceptText];
        if ( application !== null ) {
            throw new Error( 'Need To Pass application' );
        }

        const otherProperties = {
            type: 'CLOSE_APP_ALERT',
            priority: 'LOW',
            notificationType: 'alert',
            message,
            buttons,
            application,
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    SERVER_TIMED_OUT: ( appId, appName ) => {
        const title = I18n.t( 'notifications.title.server_timeout', { appName } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.retry' );
        const denyText = I18n.t(
            'notifications.buttons.denyText.cancel_install'
        );

        const otherProperties = {
            icon: 'WarningIcon',
            type: 'SERVER_TIMED_OUT',
            notificationType: 'Native',
            priority: 'HIGH',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    UPDATE_AVAILABLE: ( appId, appName, version ) => {
        const title = I18n.t( 'notifications.title.update_available', {
            appName,
            version
        } );
        const acceptText = I18n.t(
            'notifications.buttons.acceptText.update_now'
        );
        const denyText = I18n.t( 'notifications.buttons.denyText.skip' );

        const otherProperties = {
            type: 'UPDATE_AVAILABLE',
            icon: 'InfoIcon',
            priority: 'LOW',
            notificationType: 'Native',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    UPDATE_AVAILABLE_ALERT: ( appId, appName, version ) => {
        const title = I18n.t( 'notifications.title.update_available', {
            appName,
            version
        } );
        const message = I18n.t( 'notifications.message.update_available', {
            appName,
            version
        } );
        const acceptText = I18n.t(
            'notifications.buttons.acceptText.update_now'
        );
        const denyText = I18n.t( 'notifications.buttons.denyText.skip' );
        const learnMore = I18n.t( 'notifications.buttons.other.learn_more' );

        const buttons = [learnMore, denyText, acceptText];

        const otherProperties = {
            priority: 'LOW',
            type: 'UPDATE_AVAILABLE_ALERT',
            notificationType: 'alert',
            message,
            buttons,
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    ADMIN_PASS_REQ: ( appId, appName ) => {
        const title = I18n.t( 'notifications.title.admin_pass_req', { appName } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.try_again' );
        const denyText = I18n.t(
            'notifications.buttons.denyText.cancel_install'
        );

        const otherProperties = {
            type: 'ADMIN_PASS_REQ',
            priority: 'HIGH',
            icon: 'LockIcon',
            notificationType: 'Native',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    RESTART_SYSTEM: ( appName ) => {
        const title = I18n.t( 'notifications.title.restart_system', { appName } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.restart' );
        const denyText = I18n.t( 'notifications.buttons.denyText.not_now' );

        const otherProperties = {
            type: 'RESTART_SYSTEM',
            notificationType: 'Native',
            icon: 'LoopIcon',
            priority: 'LOW'
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    RESTART_SYSTEM_ALERT: ( appName ) => {
        const title = I18n.t( 'notifications.title.restart_system', { appName } );
        const message = I18n.t( 'notifications.message.restart_system', {
            appName
        } );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.restart' );
        const denyText = I18n.t( 'notifications.buttons.denyText.not_now' );
        const buttons = [denyText, acceptText];

        const otherProperties = {
            type: 'RESTART_SYSTEM_ALERT',
            notificationType: 'alert',
            priority: 'LOW',
            message,
            buttons
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    GLOBAL_FAILURE: ( appId ) => {
        const title = I18n.t( 'notifications.title.global_failure' );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.retry' );
        const denyText = I18n.t( 'notifications.buttons.denyText.dismiss' );

        const otherProperties = {
            type: 'GLOBAL_FAILURE',
            notificationType: 'Native',
            icon: 'WarningIcon',
            priority: 'HIGH',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    DISC_FULL: ( appId ) => {
        const title = I18n.t( 'notifications.title.disc_full' );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.resume' );
        const denyText = I18n.t( 'notifications.buttons.denyText.dismiss' );

        const otherProperties = {
            type: 'DISC_FULL',
            notificationType: 'Native',
            icon: 'DiscFullIcon',
            priority: 'HIGH',
            appId
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    UNDOWNLOAD_AND_INSTALL_APP_ALERT: ( appId, appName ) => {
        const title = I18n.t( 'notifications.title.uninstall_app', { appName } );
        const message = I18n.t( 'notifications.message.uninstall_app', {
            appName
        } );
        const acceptText = I18n.t(
            'notifications.buttons.acceptText.uninstall_app'
        );
        const denyText = I18n.t( 'notifications.buttons.denyText.cancel' );
        const buttons = [denyText, acceptText];

        const otherProperties = {
            type: 'UNINSTALL_APP',
            notificationType: 'alert',
            priority: 'HIGH',
            message,
            appId,
            buttons
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    },
    CLEARNET_WARNING_ALERT: () => {
        const checkboxLabel = I18n.t( 'notifications.warning.not_again' );
        const title = I18n.t( 'notifications.title.clearnet_warning' );
        const message = I18n.t( 'notifications.message.clearnet_warning' );
        const acceptText = I18n.t( 'notifications.buttons.acceptText.continue' );
        const denyText = I18n.t( 'notifications.buttons.denyText.cancel' );
        const buttons = [denyText, acceptText];

        const otherProperties = {
            type: 'CLEARNET_WARNING',
            priority: 'HIGH',
            notificationType: 'alert',
            checkboxLabel,
            message,
            buttons
        };

        return prepareNotification(
            title,
            acceptText,
            denyText,
            otherProperties
        );
    }
};
