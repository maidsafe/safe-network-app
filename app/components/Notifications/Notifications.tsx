import React, { Component } from 'react';
import { NotificationNative } from '$Components/Notifications/Notification_Native';
import { NotificationAlert } from '$Components/Notifications/Notification_Alert';
import { logger } from '$Logger';

interface Props {
    notificationCheckBox: boolean;
    notifications: Record<string, any>;
    acceptNotification: any;
    denyNotification: any;
    toggleCheckBox: any;
}

export class Notification extends React.PureComponent<Props> {
    render() {
        const {
            notifications,
            acceptNotification,
            denyNotification,
            toggleCheckBox,
            notificationCheckBox
        } = this.props;

        const notificationKeys = Object.keys( notifications );
        let l = notificationKeys.length - 1;
        if ( l >= 0 ) {
            let latestNotificationId: string = notificationKeys[l];
            let latestNotification: Record<string, any> =
                notifications[latestNotificationId];
            // eslint-disable-next-line no-plusplus
            while ( l-- && latestNotification.priority === 'LOW' ) {
                latestNotificationId = notificationKeys[l];
                latestNotification = notifications[latestNotificationId];
            }

            latestNotification.icon = latestNotification.icon || 'InfoIcon';
            latestNotification.acceptText =
                latestNotification.acceptText || 'RESUME';
            latestNotification.denyText =
                latestNotification.denyText || 'DISMISS';
            latestNotification.title =
                latestNotification.title || 'Uhh Ohh Something Went Wrong!';

            if ( latestNotification.notificationType === 'Native' ) {
                return (
                    <NotificationNative
                        latestNotification={latestNotification}
                        acceptNotification={acceptNotification}
                        denyNotification={denyNotification}
                    />
                );
            }
            if ( latestNotification.notificationType === 'alert' ) {
                return (
                    <NotificationAlert
                        latestNotification={latestNotification}
                        acceptNotification={acceptNotification}
                        denyNotification={denyNotification}
                        toggleCheckBox={toggleCheckBox}
                        notificationCheckBox={notificationCheckBox}
                    />
                );
            }
        }
        return <div />;
    }
}
