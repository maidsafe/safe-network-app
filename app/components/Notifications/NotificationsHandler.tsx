import React, { Component } from 'react';
import { Notification } from '$Components/Notifications/Notification';
import { NotificationAlert } from '$Components/Notifications/Notification_Alert';
import { logger } from '$Logger';

interface Props {
    notificationCheckBox: boolean;
    notifications: Record<string, any>;
    acceptNotification: any;
    denyNotification: any;
    toggleCheckBox: any;
}

export class NotificationsHandler extends React.PureComponent<Props> {
    render() {
        const {
            notifications,
            acceptNotification,
            denyNotification,
            toggleCheckBox,
            notificationCheckBox
        } = this.props;

        const notificationKeys = Object.keys( notifications );
        let lastNotificationIndex = notificationKeys.length - 1;
        if ( lastNotificationIndex >= 0 ) {
            let latestNotificationId: string =
                notificationKeys[lastNotificationIndex];
            let latestNotification: Record<string, any> =
                notifications[latestNotificationId];
            while (
                // eslint-disable-next-line no-plusplus
                lastNotificationIndex-- &&
                latestNotification.priority === 'LOW'
            ) {
                latestNotificationId = notificationKeys[lastNotificationIndex];
                latestNotification = notifications[latestNotificationId];
            }

            latestNotification.icon = latestNotification.icon || 'InfoIcon';
            latestNotification.acceptText =
                latestNotification.acceptText || 'RESUME';
            latestNotification.denyText =
                latestNotification.denyText || 'DISMISS';
            latestNotification.title =
                latestNotification.title || 'Uhh Ohh Something Went Wrong!';

            if ( latestNotification.notificationType === 'standard' ) {
                return (
                    <Notification
                        latestNotification={latestNotification}
                        acceptNotification={acceptNotification}
                        denyNotification={denyNotification}
                    />
                );
            }
            if ( latestNotification.notificationType === 'js-alert' ) {
                return (
                    <NotificationAlert
                        latestNotification={latestNotification}
                        acceptNotification={acceptNotification}
                        denyNotification={denyNotification}
                        toggleCheckBox={toggleCheckBox}
                        notificationCheckBox={notificationCheckBox}
                    />
                );
            }
        }
        return <div />;
    }
}
