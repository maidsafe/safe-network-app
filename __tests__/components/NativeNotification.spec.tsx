import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';
import { Button } from '@material-ui/core';
import { NotificationNative } from '$Components/Notifications/Notification_Native';
import { notificationTypes } from '$Constants/notifications';
import { logger } from '$Logger';

const shallow = createShallow();

const shallowSetup = ( propOverrides? ) => {
    const appId: string = Math.random().toString( 36 );
    const id: string = Math.random().toString( 36 );

    const props = Object.assign(
        {
            latestNotification: {
                [id]: {
                    id,
                    type: 'NO_INTERNET',
                    icon: 'SignalWifiOffIcon',
                    priority: 'HIGH',
                    notificationType: 'Native',
                    title:
                        'No Internet connection. Your install has been paused.',
                    acceptText: 'resume',
                    denyText: 'dismiss',
                    appId
                }
            },
            acceptNotification: jest.fn(),
            denyNotification: jest.fn()
        },
        propOverrides
    );

    const wrapper = shallow( <NotificationNative {...props} /> );

    return {
        props,
        wrapper
    };
};

describe( 'Notifications', () => {
    it( 'render', () => {
        const { wrapper, props } = shallowSetup();
        expect( wrapper ).toMatchSnapshot();
    } );

    it( 'accept notification button click', () => {
        const { wrapper, props } = shallowSetup();
        wrapper
            .find( Button )
            .filterWhere(
                ( node ) => node.props()['aria-label'] === 'AcceptNotification'
            )
            .simulate( 'click', { preventDefault() {} } );
        expect( props.acceptNotification ).toHaveBeenCalled();
    } );

    it( 'deny notification button click', () => {
        const { wrapper, props } = shallowSetup();
        wrapper
            .find( Button )
            .filterWhere(
                ( node ) => node.props()['aria-label'] === 'DenyNotification'
            )
            .simulate( 'click', { preventDefault() {} } );
        expect( props.denyNotification ).toHaveBeenCalled();
    } );
} );
