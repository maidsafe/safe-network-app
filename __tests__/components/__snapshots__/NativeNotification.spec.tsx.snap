// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications render 1`] = `ShallowWrapper {}`;
